src/renderer/stores/chat.js

```diff
@@ -15,85 +16,215 @@ const defaultState = {
+    }
+  } else if (type === 'SCROLLED_TO_LAST_PAGE') {
+    return { ...state, scrollToLastPage: false, scrollHeight: 0 }
+  } else if (type === 'SCROLLED_TO_BOTTOM') {
+    return { ...state, scrollToBottom: false }
+  } else if (type === 'SCROLLED_TO_BOTTOM_IF_CLOSE') {
+    return { ...state, scrollToBottomIfClose: false }
+  } else if (type === 'UI_DELETE_MESSAGE') {
+    const msgId = payload
 
-ipcRenderer.on('DD_MESSAGES_LOADED', (evt, payload) => {
-  const state = chatStore.getState()
-  const { chatId, messages, totalMessages } = payload
-  if (chatId === chatStore.getState().id) {
-    chatStore.setState({ ...state, totalMessages, messages: [...messages, ...state.messages] })
+    const messageIndex = state.messageIds.findIndex(mId => mId !== msgId)
+    const oldestFetchedMessageIndex = messageIndex === state.oldestFetchedMessageIndex
+      ? messageIndex + 1
+      : state.oldestFetchedMessageIndex
+    const messageIds = state.messageIds.filter(mId => mId !== msgId)
+    const messages = { ...state.messages, [msgId]: null }
+    return { ...state, messageIds, messages, oldestFetchedMessageIndex }
+  } else if (type === 'MESSAGE_CHANGED') {
+    return { ...state, messages: { ...state.messages, ...payload.messagesChanged } }
+  } else if (type === 'SENT_MESSAGE') {
+    const [messageId, message] = payload
+    const messageIds = [...state.messageIds, messageId]
+    const messages = { ...state.messages, [messageId]: message }
+    return { ...state, messageIds, messages, scrollToBottom: true }
+  } else if (type === 'MESSAGE_DELIVERED') {
+    const messages = {
+      ...state.messages,
+      [payload]: {
+        ...state.messages[payload],
+        msg: {
+          ...state.messages[payload].msg,
+          status: 'delivered'
+        }
+      }
+    }
+    return { ...state, messages }
   }
+  return state
 })
 
-ipcRenderer.on('DD_EVENT_MSG_UPDATE', (evt, payload) => {
-  const { chatId, messageObj, eventType } = payload
-  const state = chatStore.getState()
-  log.debug('DD_EVENT_MSG_UPDATE: ', eventType)
-  if (state.id !== chatId) {
-    return
-  }
-  if (!state.messages.find(msg => msg.id === messageObj.id)) {
-    chatStore.setState({ ...state, messages: [...state.messages, messageObj] })
-  } else {
-    const messages = state.messages.map(msg => {
-      return msg.id === messageObj.id ? messageObj : msg
+chatStore.effects.push(async ({ type, payload }, state) => {
+  if (type === 'SELECT_CHAT') {
+    const chatId = payload
+    const chat = await callDcMethodAsync('chatList.selectChat', [chatId])
+    const messageIds = await callDcMethodAsync('messageList.getMessageIds', [chatId])
+    const oldestFetchedMessageIndex = messageIds.length - PAGE_SIZE
+    const newestFetchedMessageIndex = messageIds.length
+    const messageIdsToFetch = messageIds.slice(oldestFetchedMessageIndex, newestFetchedMessageIndex)
+    const messages = await callDcMethodAsync('messageList.getMessages', [messageIdsToFetch])
```

Comment:
Why do we need 3 times to call callDcMethodAsync?
Why is messageIDs not fetched together with chatList.selectChat
Why do we need to calculate messageIdsToFetch here? PAGE_SIZE could be a shared variable, so the calculation could be done also together with chatList.selectChat